core/src/test/scala/org/typelevel/paiges/Generators.scala

```diff
@@ -93,4 +94,18 @@ object Generators {
 
   implicit val cogenDoc: Cogen[Doc] =
     Cogen[Int].contramap((d: Doc) => d.hashCode)
+
+  implicit val shrinkDoc: Shrink[Doc] = {
+    import Doc._
+    Shrink {
+      case Union(a, b) =>
+        (shrink(a) map (Union(_, b))) append (shrink(b) map (Union(a, _)))
+      case Concat(a, b) =>
+        (shrink(a) map (Concat(_, b))) append (shrink(b) map (Concat(a, _)))
```

Comment:
this could probably first try `a`and `b` by themselves: `a #:: b #:: (shrink(a) map (Concat(_, b))) append (shrink(b) map (Concat(a, _)))`

similarly above with `Union(a, b)` we can see if a or b already violates things.

also, rather than appending could we `lefts.zip(rights).flatMap { case (a, b) => a :: b :: Nil }` so that we interleave from the right and left. Otherwise scalacheck would need to exhaust all the left then the right.

cc @non who has written more of these.

Reply:
So yeah, I think in cases like concat, we should definitely try shrinking just `a` and just `b`.

Also, make sure to take a look at `interleave`: https://github.com/rickynils/scalacheck/blob/master/src/main/scala/org/scalacheck/Shrink.scala#L33 -- this allows us to try shrinking across many dimensions without having to "exhaust" any one dimension.